contracts/RMRK/utils/RMRKBulkWriterPerCollection.sol

```diff
@@ -0,0 +1,134 @@
+     * @notice Reverts if the caller is not the owner of the token
+     */
+    modifier onlyTokenOwner(uint256 tokenId) {
+        _checkTokenOwner(tokenId);
+        _;
+    }
+
+    /**
+     * @notice Initializes the contract by setting the collection
+     * @param collection Address of the collection that this contract is managing.
+     */
+    constructor(address collection) {
+        _collection = collection;
+    }
+
+    /**
+     * @notice Returns the address of the collection that this contract is managing
+     * @return Address of the collection that this contract is managing
+     */
+    function getCollection() public view returns (address) {
+        return _collection;
+    }
+
+    /**
+     * @notice Replaces the current equipped child on the asset and slot combination with the given one
+     * @dev The `IntakeEquip` stuct contains the following data:
+     *  [
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param data An `IntakeEquip` struct specifying the equip data.
+     */
+    function replaceEquip(
+        IRMRKEquippable.IntakeEquip memory data
+    ) public onlyTokenOwner(data.tokenId) {
+        IRMRKEquippable(_collection).unequip(
+            data.tokenId,
+            data.assetId,
+            data.slotPartId
+        );
+        IRMRKEquippable(_collection).equip(data);
+    }
+
+    /**
+     * @notice Performs multiple unequip and equip operations
+     * @dev Unequip operations must run first
+     * @dev tokenId is included as a parameter to be able to do a single check for ownership
+     * @dev Every tokenId in the `IntakeEquip` structs must match the tokenId passed in
+     * @dev The `IntakeUnequip` stuct contains the following data:
+     *  [
+     *      assetId,
+     *      slotPartId,
+     *  ]
+     * @dev The `IntakeEquip` stuct contains the following data:
+     *  [
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param tokenId ID of the token we are managing.
+     * @param unequips[] An array of `IntakeUnequip` structs specifying the slots to unequip.
+     * @param equips[] An array of `IntakeEquip` structs specifying the slots to equip.
+     */
+    function bulkEquip(
+        uint256 tokenId,
+        IntakeUnequip[] memory unequips,
+        IRMRKEquippable.IntakeEquip[] memory equips
+    ) public onlyTokenOwner(tokenId) {
+        uint256 length = unequips.length;
+        for (uint256 i = 0; i < length; i++) {
+            IRMRKEquippable(_collection).unequip(
+                tokenId,
+                unequips[i].assetId,
+                unequips[i].slotPartId
+            );
+        }
+        length = equips.length;
+        for (uint256 i = 0; i < length; i++) {
+            if (equips[i].tokenId != tokenId) {
+                revert RMRKCanOnlyDoBulkOperationsWithOneTokenAtATime();
```

Comment:
I think from a user standpoint, many will request/expect the ability to do bulk operations with multiple tokens.

Reply:
I initially did it that way, but that implied an external call on every step of the loop which felt really dumb.
I expect this to be called from a token detail page where you would do several operations on the same token. I don't think it would be practical for a UI to track changes from different tokens, do you?

Reply:
We can always make a 2nd version of this if need be so I guess it's fine for now.